src/socket_listener_berkeley.c

```diff
@@ -0,0 +1,169 @@
+
+typedef struct SOCKET_LISTENER_INSTANCE_TAG
+{
+	int port;
+	int socket;
+	ON_SOCKET_ACCEPTED on_socket_accepted;
+	void* callback_context;
+} SOCKET_LISTENER_INSTANCE;
+
+SOCKET_LISTENER_HANDLE socketlistener_create(int port)
+{
+	SOCKET_LISTENER_INSTANCE* result = (SOCKET_LISTENER_INSTANCE*)malloc(sizeof(SOCKET_LISTENER_INSTANCE));
+	if (result != NULL)
+	{
+		result->port = port;
+		result->on_socket_accepted = NULL;
+		result->callback_context = NULL;
+	}
+
+	return (SOCKET_LISTENER_HANDLE)result;
+}
+
+void socketlistener_destroy(SOCKET_LISTENER_HANDLE socket_listener)
+{
+	if (socket_listener != NULL)
+	{
+		socketlistener_stop(socket_listener);
+		free(socket_listener);
+	}
+}
+
+int socketlistener_start(SOCKET_LISTENER_HANDLE socket_listener, ON_SOCKET_ACCEPTED on_socket_accepted, void* callback_context)
+{
+	int result;
+
+	if (socket_listener == NULL)
+	{
+		result = __FAILURE__;
+	}
+	else
+	{
+		SOCKET_LISTENER_INSTANCE* socket_listener_instance = (SOCKET_LISTENER_INSTANCE*)socket_listener;
+
+		socket_listener_instance->socket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+		if (socket_listener_instance->socket == -1)
+		{
+			result = __FAILURE__;
+		}
+		else
+		{
+            struct sockaddr_in sa;
+
+			socket_listener_instance->on_socket_accepted = on_socket_accepted;
+			socket_listener_instance->callback_context = callback_context;
+
+            sa.sin_family = AF_INET;
+            sa.sin_port = htons(socket_listener_instance->port);
+            sa.sin_addr.s_addr = htonl(INADDR_ANY);
+
+            int flags;
+            if ((-1 == (flags = fcntl(socket_listener_instance->socket, F_GETFL, 0))) ||
+                (fcntl(socket_listener_instance->socket, F_SETFL, flags | O_NONBLOCK) == -1))
+            {
+                LogError("Failure: fcntl failure.");
+                close(socket_listener_instance->socket);
+                socket_listener_instance->socket = -1;
+                result = __FAILURE__;
+            }
+            else if (bind(socket_listener_instance->socket, (const struct sockaddr*)&sa, sizeof(sa)) == -1)
+			{
+				(void)close(socket_listener_instance->socket);
+				socket_listener_instance->socket = -1;
+				result = __FAILURE__;
+			}
+			else
+			{
+                if (listen(socket_listener_instance->socket, SOMAXCONN) == -1)
+				{
+					(void)close(socket_listener_instance->socket);
+					socket_listener_instance->socket = -1;
+					result = __FAILURE__;
+				}
+				else
+				{
+					result = 0;
+				}
+			}
+		}
+	}
+
+	return result;
+}
+
+int socketlistener_stop(SOCKET_LISTENER_HANDLE socket_listener)
+{
+	int result;
+
+	if (socket_listener == NULL)
+	{
+		result = __FAILURE__;
+	}
+	else
+	{
+		SOCKET_LISTENER_INSTANCE* socket_listener_instance = (SOCKET_LISTENER_INSTANCE*)socket_listener;
+
+		socket_listener_instance->on_socket_accepted = NULL;
+		socket_listener_instance->callback_context = NULL;
+
+		(void)close(socket_listener_instance->socket);
```

Comment:
Well, shutdown would not help here as discussed. Mybe it would be good if the IO would expose shutdown somehow. Maybe.

---
In reply to: [109061082](https://github.com/Azure/azure-uamqp-c/pull/69#discussion_r109061082) [](ancestors = 109061082)